### PR TITLE
Interpolation to and from sub-meshes for co-dimension 0

### PR DESCRIPTION
**Pull request by jorgensd**
A feature that is quite neat for post-processing, as one often only wants to visualize on a sub mesh.
Can also be used to avoid mixed-assembly in cases of iterative schemes:
1. Solve on a parent mesh
2. Move solution to sub mesh
3. Solve equation on sub mesh
4. Move solution back and repeat.

There is a slight API change in the C++ interface, as expressions need the associated mesh to be supplied.
In Python, the API only changes if you want to map between meshes.

### CODE REVIEW

**Review by jhale — APPROVED** [review]

File: cpp/dolfinx/fem/Function.h [hunk 1]
```diff
@@ -331,8 +340,28 @@ class Function
...
+    std::vector<std::int32_t> cells_expr;
+    cells_expr.reserve(num_cells);
+    // Get mesh and check if mesh is shared
+    if (auto mesh_v = _function_space->mesh();
+        expr_mesh.topology() == mesh_v->topology())
+    {
+      cells_expr.insert(cells_expr.end(), cells.begin(), cells.end());
+    }
+    // If meshes are different and input mapping is given
+    else if (cell_map.size() > 0)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Function.h:354]:
```suggestion
    else if (!cell_map.empty())
```

File: cpp/dolfinx/fem/Function.h [hunk 2]
```diff
@@ -356,17 +385,27 @@ class Function
 
   /// Interpolate an Expression (based on UFL) on all cells
   /// @param[in] e The function to be interpolated
-  void interpolate(const Expression<value_type, geometry_type>& e)
+  /// @param[in] expr_mesh Mesh expression is defined on
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Function.h:388]:
```suggestion
  /// @param[in] mesh Mesh the Expression `e` is defined on.
```

File: cpp/dolfinx/fem/Function.h [hunk 3]
```diff
@@ -356,17 +385,27 @@ class Function
...
+  /// @param[in] cell_map Map from `cells` to cells in expression if
+  /// receiving function is defined on a different mesh than the expression
+  void interpolate(const Expression<value_type, geometry_type>& e,
+                   const dolfinx::mesh::Mesh<geometry_type>& expr_mesh,
+                   std::span<const std::int32_t> cell_map
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Function.h:393]: See previous comment of default args.

File: cpp/dolfinx/fem/interpolate.h [hunk 4]
```diff
@@ -410,37 +418,45 @@ void interpolate_same_map(Function<T, U>& u1, const Function<T, U>& u0,
 /// standard isoparametric mapping.
 /// @param[out] u1 The function to interpolate to
 /// @param[in] u0 The function to interpolate from
-/// @param[in] cells The cells to interpolate on
+/// @param[in] cells_u1 The cells to interpolate on
+/// @param[in] cells_u0 Equivalent cell in u0 for each cell in u1
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:422]: This could be improved. Short example in the docstring would also help.

File: cpp/dolfinx/fem/interpolate.h [hunk 5]
```diff
@@ -410,37 +418,45 @@ void interpolate_same_map(Function<T, U>& u1, const Function<T, U>& u0,
...
 /// @pre The functions `u1` and `u0` must share the same mesh. This is
 /// not checked by the function.
 template <dolfinx::scalar T, std::floating_point U>
 void interpolate_nonmatching_maps(Function<T, U>& u1, const Function<T, U>& u0,
-                                  std::span<const std::int32_t> cells)
+                                  std::span<const std::int32_t> cells_u1,
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:427]: Names are clunky - would `cells1` and `cells0` make sense?

File: cpp/dolfinx/fem/interpolate.h [hunk 6]
```diff
@@ -1097,14 +1113,16 @@ create_nonmatching_meshes_interpolation_data(const mesh::Mesh<T>& mesh0,
 /// @brief Interpolate from one finite element Function to another one.
 /// @param[out] u The function to interpolate into
 /// @param[in] v The function to be interpolated
-/// @param[in] cells List of cell indices to interpolate on
+/// @param[in] cells_u List of cell indices to interpolate on
+/// @param[in] cell_map Mapping from cells in `u` to cells in `v`
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:1117]: Docstring isn't clear.

File: cpp/dolfinx/fem/interpolate.h [hunk 7]
```diff
@@ -1097,14 +1113,16 @@ create_nonmatching_meshes_interpolation_data(const mesh::Mesh<T>& mesh0,
...
 /// @param[in] nmm_interpolation_data Auxiliary data to interpolate on
 /// nonmatching meshes. This data can be generated with
 /// create_nonmatching_meshes_interpolation_data (optional).
 template <dolfinx::scalar T, std::floating_point U>
 void interpolate(
     Function<T, U>& u, const Function<T, U>& v,
-    std::span<const std::int32_t> cells,
+    std::span<const std::int32_t> cells_u,
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:1124]: Better names?

File: cpp/dolfinx/fem/interpolate.h [hunk 8]
```diff
@@ -1127,79 +1145,91 @@ void interpolate(
...
-    else
+    // If meshes are different and input mapping is given
+    else if (cell_map.size() > 0)
+    {
+      std::transform(cells_u.begin(), cells_u.end(),
+                     std::back_inserter(cells_v),
+                     [&cell_map](std::int32_t c) { return cell_map[c]; });
+    }
+    // Non-matching meshes
+    if (cells_v.size() == 0)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:1163]:
```suggestion
    if (cells_v.empty())
```

File: python/dolfinx/wrappers/mesh.cpp [hunk 9]
```diff
@@ -301,8 +301,12 @@ void declare_mesh(nb::module_& m, std::string type)
       [](const dolfinx::mesh::Mesh<T>& mesh, int dim,
          nb::ndarray<const std::int32_t, nb::ndim<1>, nb::c_contig> entities)
       {
-        return dolfinx::mesh::create_submesh(
+        auto [submesh, e_map, v_map, g_map] = dolfinx::mesh::create_submesh(
             mesh, dim, std::span(entities.data(), entities.size()));
+        auto _e_map = as_nbarray(e_map);
```
Comment (garth-wells) [inline python/dolfinx/wrappers/mesh.cpp:306]: I think this will mean arrays are created three times. It could be restructured to use move semantics.

File: python/test/unit/fem/test_interpolation.py [hunk 10]
```diff
@@ -1025,3 +1026,109 @@ def f_test2(x):
 
     l2_error = assemble_scalar(form((u1 - u1_exact) ** 2 * dx_cell(cell_label), dtype=xtype))
     assert np.isclose(l2_error, 0.0, rtol=np.finfo(xtype).eps, atol=np.finfo(xtype).eps)
+
+
+def test_submesh_interpolation():
+    mesh = create_unit_square(MPI.COMM_WORLD, 6, 7)
```
Comment (garth-wells) [inline python/test/unit/fem/test_interpolation.py:1032]: Add short doc string.

File: python/test/unit/fem/test_interpolation.py [hunk 11]
```diff
@@ -1025,3 +1026,109 @@ def f_test2(x):
...
+    w.interpolate(u_sub_exact, cells=sub_to_parent, cell_map=parent_to_sub)
+
+    w_exact = Function(W)
+    w_exact.interpolate(ref_func, cells=cells)
+
+    np.testing.assert_allclose(w.x.array, w_exact.x.array, atol=atol)
+
+
+def test_submesh_expression_interpolation():
+    mesh = create_unit_square(MPI.COMM_WORLD, 10, 8, cell_type=CellType.quadrilateral)
```
Comment (garth-wells) [inline python/test/unit/fem/test_interpolation.py:1078]: Add short docstring explaining what test does.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/interpolate.h (hunk 7 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/interpolate.h:1124]: Do you have a suggestion? We need to make it clear that the cells sent in are associated with the mesh of `u`, while the map maps each cell in `u` to a cell in `v`.

Comment (garth-wells) [reply cpp/dolfinx/fem/interpolate.h:1125]: Use `u0` and `u1` for the functions, the `cells0` etc.

**Comment by jorgensd** [comment]
@garth-wells i’ve tried to address your comments.

File: cpp/dolfinx/fem/Function.h [hunk 12]
```diff
@@ -331,8 +341,26 @@ class Function
 
     // Evaluate Expression at points
     assert(_function_space->mesh());
-    e.eval(*_function_space->mesh(), cells, fdata,
-           {num_cells, num_points * value_size});
+
+    //
```
Comment (jorgensd) [inline cpp/dolfinx/fem/Function.h:345]:
```suggestion
```

File: python/dolfinx/fem/function.py [hunk 13]
```diff
@@ -420,6 +420,8 @@ def interpolate(
         self,
         u: typing.Union[typing.Callable, Expression, Function],
         cells: typing.Optional[np.ndarray] = None,
+        cell_map: typing.Optional[np.ndarray] = None,
+        expr_mesh: typing.Optional[Mesh] = None,
```
Comment (jpdean) [inline python/dolfinx/fem/function.py:424]: `expr_mesh` isn't documented. It would be cleaner if there was a way to avoid passing it

File: python/dolfinx/fem/function.py [hunk 14]
```diff
@@ -428,6 +430,7 @@ def interpolate(
             u: The function, Expression or Function to interpolate.
             cells: The cells to interpolate over. If `None` then all
                 cells are interpolated over.
+            cell_map: Mapping from cells in receiving mesh to cells in `u` mesh
```
Comment (jpdean) [inline python/dolfinx/fem/function.py:433]: What's the "receiving mesh"? Presumably `cell_map` maps `cells` to the cells in the `u` mesh?

File: python/dolfinx/fem/function.py [hunk 15]
```diff
@@ -456,13 +467,26 @@ def _(u_ptr: int, cells: typing.Optional[np.ndarray] = None):
...
+            Args:
+                expr: Expression to interpolate
+                cells: The cells to interpolate over. If `None` then all
+                    cells are interpolated over.
+            """
+            assert cell_map is not None
+            if len(cell_map) == 0:
+                # If cell map is not provided create identity map
+                assert cells is not None
+                expr_cell_map = np.arange(len(cells), dtype=np.int32)
```
Comment (jpdean) [inline python/dolfinx/fem/function.py:480]: Would it be better to pass the cells in the other mesh directly? This would mean we don't need to create an identity map (you could just pass `cells` here instead). We might also not need to pass the expression mesh as well in this case

**Review by jpdean — COMMENTED** [review]
I think it's a cool feature! :) Here are a few initial comments. I'll have a look at the rest soon

File: python/dolfinx/fem/function.py (hunk 14 above)
Comment (jorgensd) [inline python/dolfinx/fem/function.py:433]:
```suggestion
            cell_map: Mapping from `cells` to to cells in the mesh that `u` is defined over.
```

Comment (jorgensd) [reply python/dolfinx/fem/function.py:433]: @jpdean how about this?

File: python/test/unit/fem/test_interpolation.py [hunk 16]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
 
     l2_error = assemble_scalar(form((u1 - u1_exact) ** 2 * dx_cell(cell_label), dtype=xtype))
     assert np.isclose(l2_error, 0.0, rtol=np.finfo(xtype).eps, atol=np.finfo(xtype).eps)
+
+
+def test_submesh_interpolation():
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1031]: It goes both ways?

File: python/dolfinx/fem/function.py (hunk 13 above)
Comment (jorgensd) [reply python/dolfinx/fem/function.py:424]: I don't see any way to do this, as an expression is not tied to a `Mesh`, after https://github.com/FEniCS/dolfinx/pull/2730

File: python/dolfinx/fem/function.py (hunk 15 above)
Comment (jorgensd) [reply python/dolfinx/fem/function.py:480]: Mesh is required due to the current way expressions are implemented (see other comment).
I thought about just supplying the cells here, but it would mean more work has to be done in Python by the user.

File: python/dolfinx/fem/function.py (hunk 13 above)
Comment (jpdean) [reply python/dolfinx/fem/function.py:424]: Ah OK, in that case `expr_mesh` just needs documenting then

File: cpp/dolfinx/fem/Function.h [hunk 17]
```diff
@@ -153,28 +153,33 @@ class Function
   /// @brief Interpolate a provided Function.
   /// @param[in] v The function to be interpolated
   /// @param[in] cells The cells to interpolate on
+  /// @param[in] cell_map Map from indices in `cells` to cell indices in `v`
```
Comment (jpdean) [inline cpp/dolfinx/fem/Function.h:156]: Maybe this is clearer: `cell_map` A map from cells in the mesh associated with `this` function to cells in mesh associated with `v`

File: cpp/dolfinx/fem/Function.h [hunk 18]
```diff
@@ -283,8 +288,13 @@ class Function
   /// `FiniteElement::interpolation_points()` for the element associated
   /// with `u`.
   /// @param[in] cells The cells to interpolate on
+  /// @param[in] expr_mesh Mesh expression is defined on
```
Comment (jpdean) [inline cpp/dolfinx/fem/Function.h:291]:
```suggestion
  /// @param[in] expr_mesh The mesh to evaluate the expression on
```

File: cpp/dolfinx/fem/Function.h (hunk 17 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/Function.h:156]:
```suggestion
  /// @param[in] cell_map A map from cells in the mesh associated with `this` function to cells in mesh associated with `v`
```

**Comment by jhale** [comment]
This is very useful and I'd like to see it go into 0.8. Please merge this once you are happy with it.

File: python/test/unit/fem/test_interpolation.py [hunk 19]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
...
+        values = np.zeros((2, x.shape[1]), dtype=default_scalar_type)
+        values[0] = -6 * x[0]
+        values[1] = 2 * x[1]
+        return values
+
+    def grad_squared(x):
+        grad = grad_ref_func(x)
+        return grad[0] ** 2 + grad[1] ** 2
+
+    tdim = mesh.topology.dim
```
Comment (jhale) [inline python/test/unit/fem/test_interpolation.py:1098]: Does this work for the case when `tdim` is not the same between parent and submesh?

File: python/test/unit/fem/test_interpolation.py [hunk 20]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
...
+    parent_to_sub[sub_to_parent] = np.arange(len(sub_to_parent))
+
+    sub_expr = Expression(ufl.dot(u_sub_exact, u_sub_exact), W.element.interpolation_points())
+
+    # Mapping back needs to be restricted to the subset of cells in the submesh
+    w.interpolate(sub_expr, cells=sub_to_parent, expr_mesh=submesh, cell_map=parent_to_sub)
+
+    w_exact = Function(W)
+    w_exact.interpolate(grad_squared, cells=cells)
+    np.testing.assert_allclose(w.x.array, w_exact.x.array, atol=5e-4)
```
Comment (jhale) [inline python/test/unit/fem/test_interpolation.py:1136]: Tolerance seems quite high. Why?

File: python/test/unit/fem/test_interpolation.py [hunk 21]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
...
+
+    np.testing.assert_allclose(w.x.array, w_exact.x.array, atol=atol)
+
+
+def test_submesh_expression_interpolation():
+    """Test interpolation of an expression between a submesh and its parent"""
+    mesh = create_unit_square(MPI.COMM_WORLD, 10, 8, cell_type=CellType.quadrilateral)
+
+    def left_locator(x):
+        return x[0] <= 0.5 + 1e-14
```
Comment (jhale) [inline python/test/unit/fem/test_interpolation.py:1083]: Not in house style `np.isclose`?

File: python/test/unit/fem/test_interpolation.py [hunk 22]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
...
     l2_error = assemble_scalar(form((u1 - u1_exact) ** 2 * dx_cell(cell_label), dtype=xtype))
     assert np.isclose(l2_error, 0.0, rtol=np.finfo(xtype).eps, atol=np.finfo(xtype).eps)
+
+
+def test_submesh_interpolation():
+    """Test interpolation of a function between a submesh and its parent"""
+    mesh = create_unit_square(MPI.COMM_WORLD, 6, 7)
+
+    def left_locator(x):
+        return x[0] <= 0.5 + 1e-14
```
Comment (jhale) [inline python/test/unit/fem/test_interpolation.py:1036]: Not in house style `np.isclose`

File: python/test/unit/fem/test_interpolation.py [hunk 23]
```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
...
+    atol = 5 * np.finfo(default_scalar_type).resolution
+    np.testing.assert_allclose(u_sub_exact.x.array, u_sub.x.array, atol=atol)
+
+    # Map from sub to parent
+    W = functionspace(mesh, ("DQ", 2))
+    w = Function(W)
+
+    cell_imap = mesh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    parent_to_sub = np.full(num_cells, -1, dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1126]: As a parent mesh has more cells than the submesh has, we need to indicate this somehow (i.e. that a parent cell is not in the submesh). This is how we indicate it throughout the code when it comes to cell indices.

File: python/test/unit/fem/test_interpolation.py (hunk 19 above)
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1098]: No, it does not. I've shown that this in theory could be possible through: https://fenicsproject.discourse.group/t/setting-scalar-function-for-bdm-boundary-condition-with-controlled-flux-mixed-poisson/13508/9?u=dokken
But it would require much more control over the interpolation points in `basix`.

File: python/test/unit/fem/test_interpolation.py (hunk 21 above)
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1083]: It is not just values close to 0.5, it is all cells with vertices smaller or equal to 0.5.

File: python/test/unit/fem/test_interpolation.py (hunk 22 above)
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1036]: Same comment as below, it is all values smaller or equal to 0.5

File: python/test/unit/fem/test_interpolation.py (hunk 20 above)
Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py:1136]: Honestly, I do not know.

Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py]: Stricter tolerances fail single precision checks: 
https://github.com/FEniCS/dolfinx/actions/runs/8787873760/job/24114119650?pr=3114
Note that the difference is very small:
```bash
E           Mismatched elements: 79 / 720 (11%)
E           Max absolute difference: 1.8119812e-05
E           Max relative difference: 0.52334005
E            x: array([4.024558e-16, 6.250000e-02, 3.600000e-01, 4.224998e-01,
E                  1.562500e-02, 8.999999e-02, 1.524999e-01, 3.756249e-01,
E                  1.056250e-01, 6.249999e-02, 2.500000e-01, 4.224999e-01,...
E            y: array([8.443247e-16, 6.250000e-02, 3.600000e-01, 4.225002e-01,
E                  1.562500e-02, 9.000000e-02, 1.525001e-01, 3.756251e-01,
E                  1.056250e-01, 6.249999e-02, 2.500000e-01, 4.225000e-01,...

```
Seems like there is some floating point issue in the evaluation?

Comment (jorgensd) [reply python/test/unit/fem/test_interpolation.py]: There seems to be some floating error accumulation over multiple steps here. I have revised the test to avoid floating point error accummulation

**Comment by jorgensd** [comment]
All tests now pass. I've tried to address all comments.

File: python/test/unit/fem/test_interpolation.py [hunk 24]
```diff
@@ -1025,3 +1026,114 @@ def f_test2(x):
...
+    u.interpolate(ref_func)
+
+    V_sub = functionspace(submesh, ("N2curl", 1))
+    u_sub = Function(V_sub)
+
+    parent_expr = Expression(ufl.grad(u), V_sub.element.interpolation_points())
+
+    # Map from parent to sub mesh
+
+    u_sub.interpolate(parent_expr, expr_mesh=mesh, cell_map=sub_to_parent)
```
Comment (garth-wells) [inline python/test/unit/fem/test_interpolation.py:1113]: The CI deadlock https://github.com/FEniCS/dolfinx/actions/runs/8838261829 is triggered by this line.
